### PR TITLE
CCIP-2003 Improved fetching Commit Reports from database 

### PR DESCRIPTION
**Pull request by mateusz-sekara**
## Motivation

Current implementation fetches all the CommitReports within the permissionlessExecThreshold from database and filter out result set based on the `snoozedRoots`. It means that when everything is executed, we keep fetching all roots to memory only to discover that we don't need that data. During high traffic spikes, this is an unnecessary waste of resources in the CL node and database as the dataset size grows in time till it reaches permissionlessExec window (8 hours of logs for prod).

Example load run in which you can see how it changes over time during steady traffic of messages

image

## Solution

There are at least two solutions to make it more performant

1. Pass executed roots to the database query to fetch only unexpired roots from the database. Unexpired roots can be determined from the snoozedRoots cache. This is probably the most accurate approach because it always fetches the minimal required set from the database.

You can think of the query like this
```sql
select * from evm.logs
where address = :addr
and event_sig = :event_sig
and substring(data, X, Y) not in (:executed_roots_array)
and block_timestamp > :permissionlessThresholdWindow
```

2. Make permissionlessThreshold filter "sliding". Commit Reports are executed sequentially for most of the cases, so we can keep track of timestamps of the oldest fully executed report. Instead of inventing a new query for LogPoller we can achieve a lot based on the assumption that execution is sequential. Whenever report is markedAsExecuted we keep bumping block_timestamp filter. We already have that information in cache and indirectly rely on that when filtering out snoozedRoots
```sql
and block_timestamp > :oldest_executed_timestamp 
```

Solution 2 is easier to implement, doesn't require any changes to LogPoller (and query perf testing) and can be handled in the plugin. We already cache snoozedRoots so we can reuse that to move filtering to the database by narrowing the block_timestamp search window. Assuming that reports are executed sequentially, it should be as accurate as solution 1.

### Load testing solution 2 (the same configuration, but shorter test duration)
image

### CODE REVIEW

File: core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	// CleanupInterval defines how often roots cache is scanned to evict stale roots
+	CleanupInterval = 30 * time.Minute
+)
+
+type CommitsRootsCache interface {
+	IsSnoozed(merkleRoot [32]byte) bool
+	MarkAsExecuted(merkleRoot [32]byte)
+	Snooze(merkleRoot [32]byte)
+
+	CommitSearchTimestamp() time.Time
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go]: It's kind of impossible to understand what this function is doing without reading the code.

File: core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	// Example
+	// permissionLessExecThresholds - 10 days, now - 2010-10-15
+	// We fetch all the roots that within the [2010-10-05, 2010-10-15] window and load them to the queue
+	// [0xA - 2010-10-10, 0xB - 2010-10-11, 0xC - 2010-10-12] -> 0xA is the oldest root
+	// We executed 0xA and a couple of rounds later, we mark 0xA as executed and snoozed that forever which removes it from the queue.
+	// [0xB - 2010-10-11, 0xC - 2010-10-12]
+	// Now the search filter wil be 0xA timestamp -> [2010-10-11, 20-10-15]
+	// If roots are executed out of order, it's not going to change anything. However, for most of the cases we have sequential root execution and that is
+	// a huge improvement because we don't need to fetch all the roots from the database in every round.
+	unexecutedRootsQueue *orderedmap.OrderedMap[string, time.Time]
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go:59]: I would prefer not to use [this lib](https://github.com/wk8/go-ordered-map).
We could have our own simple implementation or use a different type.

Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go:59]: If we have our own simple implementation you can also have the `rootsQueueMu` defined there.

File: core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+		rootSnoozeTime,
+		EvictionGracePeriod,
+		CleanupInterval,
+	)
+}
+
+func (s *commitRootsCache) IsSnoozed(merkleRoot [32]byte) bool {
+	_, snoozed := s.snoozedRoots.Get(merkleRootToString(merkleRoot))
+	_, executed := s.executedRoots.Get(merkleRootToString(merkleRoot))
+	return snoozed || executed
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go:106]: The logic seems weird. `IsSnoozed = Snoozed or Executed`.
```
interface {
   IsSnoozed()
   MarkAsSnoozed()
   MarkAsExecuted()
}
```

e.g. A client can `MarkAsExecuted` and then call `IsSnoozed` which returns `true` but he never `snoozed` it.
We can let the client do the `OR` e.g. `if c.IsSnoozed() or c.IsExecuted() { ... }` or rename it `IsSNoozedOrExecuted`

File: core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+func (s *commitRootsCache) Snooze(merkleRoot [32]byte) {
+	s.snoozedRoots.SetDefault(merkleRootToString(merkleRoot), struct{}{})
+}
+
+func (s *commitRootsCache) CommitSearchTimestamp() time.Time {
+	permissionlessExecWindow := time.Now().Add(-s.permissionLessExecutionThresholdDuration)
+
+	timestamp, ok := func() (time.Time, bool) {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go]: should be cleaner to just place it in a separate method

File: core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	// permissionLessExecThresholds - 10 days, now - 2010-10-15
+	// We fetch all the roots that within the [2010-10-05, 2010-10-15] window and load them to the queue
+	// [0xA - 2010-10-10, 0xB - 2010-10-11, 0xC - 2010-10-12] -> 0xA is the oldest root
+	// We executed 0xA and a couple of rounds later, we mark 0xA as executed and snoozed that forever which removes it from the queue.
+	// [0xB - 2010-10-11, 0xC - 2010-10-12]
+	// Now the search filter wil be 0xA timestamp -> [2010-10-11, 20-10-15]
+	// If roots are executed out of order, it's not going to change anything. However, for most of the cases we have sequential root execution and that is
+	// a huge improvement because we don't need to fetch all the roots from the database in every round.
+	unexecutedRootsQueue *orderedmap.OrderedMap[string, time.Time]
+	rootSearchFilter     time.Time
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go]: `rootSearchFilter` is a bit tricky to understand what it does

File: core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+	if ok {
+		return timestamp
+	}
+
+	s.rootsQueueMu.Lock()
+	defer s.rootsQueueMu.Unlock()
+
+	// If rootsSearchFilter is before permissionlessExecWindow, it means that we have roots that are stuck forever and will never be executed
+	// In that case, we wipe out the entire queue. Next round should start from the permissionlessExecThreshold and rebuild cache from the scratch.
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go]: `from the scratch` -> `from scratch`

File: core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go (hunk 2 above)
Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go:59]: It was already in the deps so I've decided to reuse it. Do you have any better ideas? I'm not sure about reimplementing, sounds like reinventing the wheel.

File: core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go (hunk 5 above)
Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go]: How about `oldestRootTimestamp`?

File: core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go (hunk 3 above)
Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go:106]: Renamed to `IsSkipped` and `IsSkipped = executed or snoozed`

File: core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go (hunk 2 above)
Comment (mateusz-sekara) [reply core/services/ocr2/plugins/ccip/internal/cache/commit_roots.go:59]: Alternative implementation would be this https://github.com/elliotchance/orderedmap (more GH stars)

**Review by dimkouv — APPROVED** [review]

**Pull request merged**